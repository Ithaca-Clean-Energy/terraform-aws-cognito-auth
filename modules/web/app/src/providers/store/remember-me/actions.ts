/*
 * Copyright (c) 2018-2019 Martin Donath <martin.donath@squidfunk.com>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to
 * deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 * IN THE SOFTWARE.
 */

import { Action } from "redux"

/* ----------------------------------------------------------------------------
 * Types
 * ------------------------------------------------------------------------- */

/**
 * Remember me action types
 */
export enum RememberMeActionTypes {
  SET = "REMEMBER_ME_SET",
  SET_RESULT = "REMEMBER_ME_SET_RESULT"
}

/* ------------------------------------------------------------------------- */

/**
 * Remember me set action
 */
interface RememberMeSetAction extends Action {
  type: RememberMeActionTypes.SET
  active: boolean
}

/**
 * Remember me set result action
 */
interface RememberMeSetResultAction extends Action {
  type: RememberMeActionTypes.SET_RESULT
  result: boolean
}

/* ------------------------------------------------------------------------- */

/**
 * Remember me actions
 */
export type RememberMeActions =
  | RememberMeSetAction
  | RememberMeSetResultAction

/* ----------------------------------------------------------------------------
 * Actions
 * ------------------------------------------------------------------------- */

/**
 * Set remember me flag action creator
 *
 * @param active - Whether to remember the user
 *
 * @return Action
 */
export const setRememberMeAction =
  (active: boolean): RememberMeSetAction => ({
    type: RememberMeActionTypes.SET, active
  })

/**
 * Set authentication result action creator
 *
 * @param result - Authentication result
 *
 * @return Action
 */
export const setRememberMeResultAction =
  (result: boolean): RememberMeSetResultAction => ({
    type: RememberMeActionTypes.SET_RESULT, result
  })
