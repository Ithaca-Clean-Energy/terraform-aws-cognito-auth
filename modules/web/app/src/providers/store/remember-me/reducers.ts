/*
 * Copyright (c) 2018 Martin Donath <martin.donath@squidfunk.com>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to
 * deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 * IN THE SOFTWARE.
 */

import {
  RememberMeActions,
  RememberMeActionTypes
} from "./actions"

/* ----------------------------------------------------------------------------
 * Types
 * ------------------------------------------------------------------------- */

/**
 * Remember me state
 *
 * The refresh state needs to be persisted across sessions to properly handle
 * token-based authentication as the refresh token is persisted as a secure
 * HTTP-only cookie, and thus cannot be read from the application.
 *
 * If `failed` is set to `false`, the application will authenticate during the
 * initial load in the background in order to refresh the session without
 * prompting for credentials again. If this fails, the application will set the
 * value to `true`, effectively disabling token-based authentication until the
 * next credential-based authentication.
 *
 * The `active` value determines the checked state of the remember me checkbox,
 * so the user doesn't have to re-check it everytime the refresh token expires
 * and he has to authenticate again.
 */
export interface RememberMeState {
  active: boolean                      /* Whether to remember the user */
  failed: boolean                      /* Whether token-based */
}

/* ----------------------------------------------------------------------------
 * Reducer
 * ------------------------------------------------------------------------- */

/**
 * Remember me reducer
 *
 * @param state - Remember me state
 * @param action - Remember me action
 */
export function remember(
  state: RememberMeState = { active: false, failed: true },
  action: RememberMeActions
): RememberMeState {
  switch (action.type) {
    case RememberMeActionTypes.SET:
      return { ...state, active: action.active }
    case RememberMeActionTypes.FAILED:
      return { ...state, failed: action.failed }
    default:
      return state
  }
}
