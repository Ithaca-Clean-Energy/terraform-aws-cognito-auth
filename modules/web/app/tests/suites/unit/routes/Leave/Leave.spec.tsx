/*
 * Copyright (c) 2018 Martin Donath <martin.donath@squidfunk.com>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to
 * deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 * IN THE SOFTWARE.
 */

import { mount, shallow } from "enzyme"
import * as React from "react"
import { MemoryRouter } from "react-router"

import { setRememberMeResultAction } from "providers/store/remember-me"
import {
  SessionActionTypes,
  terminateSessionAction
} from "providers/store/session/actions"
import {
  Leave,
  Render
} from "routes/Leave/Leave"

import { search } from "_/helpers"
import { mockStore } from "_/mocks/providers"
import {
  mockAxiosPostWithError,
  mockAxiosPostWithResult
} from "_/mocks/vendor/axios"

/* ----------------------------------------------------------------------------
 * Tests
 * ------------------------------------------------------------------------- */

/* Sign out component */
describe("components/Leave", () => {

  /* Render component */
  describe("Render", () => {

    /* Test: should render with default props */
    it("should render with default props", () => {
      const wrapper = shallow(<Render />)
      expect(wrapper).toMatchSnapshot()
    })
  })

  /* Enhanced component */
  describe("Leave", () => {

    /* Initialize store */
    const store = mockStore({
      remember: {
        active: false
      }
    })

    /* Clear store */
    beforeEach(() => {
      store.clearActions()
    })

    /* Test: should render with default props */
    it("should render with default props", () => {
      mockAxiosPostWithResult()
      const wrapper = shallow(<Leave />, {
        context: { store }
      })
      expect(search(wrapper, Render)).toMatchSnapshot()
    })

    /* Test: should disable token-based authentication before mount */
    it("should disable token-based authentication before mount", () => {
      mockAxiosPostWithResult()
      mount(
        <MemoryRouter>
          <Leave />
        </MemoryRouter>, {
          context: { store },
          childContextTypes: {
            store: () => null
          }
        }
      )
      expect(store.getActions().length).toEqual(2)
      expect(store.getActions()[0])
        .toEqual(setRememberMeResultAction(false))
    })

    /* Test: should send request form after mount */
    it("should send request after mount", () => {
      const postMock = mockAxiosPostWithResult()
      mount(
        <MemoryRouter>
          <Leave />
        </MemoryRouter>, {
          context: { store },
          childContextTypes: {
            store: () => null
          }
        }
      )
      expect(postMock).toHaveBeenCalled()
    })

    /* Test: should terminate session for successful request */
    it("should terminate session for successful request", async () => {
      mockAxiosPostWithResult()
      mount(
        <MemoryRouter>
          <Leave />
        </MemoryRouter>, {
          context: { store },
          childContextTypes: {
            store: () => null
          }
        }
      )
      await new Promise(resolve => setTimeout(resolve, 500))
      expect(store.getActions().find(action => {
        return action.type === SessionActionTypes.TERMINATE
      }))
        .toEqual(terminateSessionAction())
    })

    /* Test: should terminate session for failed request */
    it("should terminate session for failed request", async () => {
      mockAxiosPostWithError()
      mount(
        <MemoryRouter>
          <Leave />
        </MemoryRouter>, {
          context: { store },
          childContextTypes: {
            store: () => null
          }
        }
      )
      await new Promise(resolve => setTimeout(resolve, 500))
      expect(store.getActions().find(action => {
        return action.type === SessionActionTypes.TERMINATE
      }))
        .toEqual(terminateSessionAction())
    })
  })
})
