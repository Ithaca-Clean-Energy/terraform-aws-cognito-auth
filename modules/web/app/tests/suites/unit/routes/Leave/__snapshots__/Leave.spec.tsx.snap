// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Leave Leave should render with default props 1`] = `<Leave />`;

exports[`components/Leave Render should render with default props 1`] = `<pure(Loading) />`;
